engine/crates/partial-caching/src/fetching/keys.rs

```diff
@@ -0,0 +1,115 @@
+use std::collections::BTreeSet;
+
+use common_types::auth::ExecutionAuth;
+use engine_value::{ConstValue, Variables};
+use registry_for_cache::{CacheAccessScope, CacheControl};
+
+use super::CachingPlan;
+use crate::query_subset::QuerySubsetDisplay;
+
+pub fn build_cache_keys(
+    plan: &CachingPlan,
+    auth: &ExecutionAuth,
+    headers: &http::HeaderMap,
+    variables: &Variables,
+) -> Vec<Option<String>> {
+    plan.cache_partitions
+        .iter()
+        .map(|(cache_control, query_subset)| {
+            // If we can't figure out the scope for a cache_control entry we just skip it.
+            // This will force it to always be fetched from the cache.
+            let scopes = cache_scopes(cache_control, auth, headers)?;
+
+            let serializer = CacheKeySerializer {
+                query: query_subset.as_display(&plan.document),
+                scopes,
+                variables: variables_required(query_subset, &plan.document, variables),
+            };
+
+            match serializer.build_key_string() {
+                Ok(key) => Some(key),
+                Err(err) => {
+                    tracing::error!("Could not build key string for {cache_control:?}: {err:?}");
+                    None
+                }
+            }
+        })
+        .collect()
+}
+
+fn cache_scopes<'a>(
+    cache_control: &CacheControl,
+    auth: &'a ExecutionAuth,
+    headers: &http::HeaderMap,
+) -> Option<CacheAccess<'a>> {
+    let Some(scopes) = &cache_control.access_scopes else {
```

Comment:
IMHO this behavior seems useless to me. If the whole token is the scope, I understand cache control is private and should only be cached by the browse, but well that's what we did before. So might be worth a comment, or just ignore me 🤷

Reply:
I don't think this is useless.  

If there's some really expensive field, and it appears in a whole range of queries - why wouldn't we want to cache that, even if the cache is user specific?  Sure, it could be cached by the client, but as soon as the user opens a tab or hits refresh that cache is gone.  Or maybe it's a CLI that's running the call which doesn't even have a cache.

Or if the auth token is shared amongst many users or many devices.

Reply:
Well for partial caching it's indeed a bit different.

> Sure, it could be cached by the client, but as soon as the user opens a tab or hits refresh that cache is gone.

Would still be there, in the browser cache. Unless I don't understand at all how it works. But on another device no indeed. I doubt the same token would be used across devices though. 

The other problem with caching potentially private stuff is we don't know what's in there and how it'll be stored by CF. With `Cache-Control: private` header a server should not cache the response at all and only let the client should do it, scoping by auth feels like it but we don't behave in the same way, so well. Not convinced it's a good default behavior.

Reply:
> Would still be there, in the browser cache

I don't know that it would.  Any headers we return to the client need to use the minimum cache age of any individual field in the response.  So if this theoretical field is cacheable for 60 days, but everything else in the response is only cacheable for 30 seconds, then after 30 seconds the browser has no cache.

> I doubt the same token would be used across devices though.

In a lot of cases sure, but it's not impossible.  

Also bear in mind that ExecutionAuth isn't just tokens - if it's public auth then that goes in there too.  So we get a separate cache for public users, which seems faintly sensible.

> The other problem with caching potentially private stuff is we don't know what's in there and how it'll be stored by CF. With Cache-Control: private header a server should not cache the response at all and only let the client should do it, scoping by auth feels like it but we don't behave in the same way, so well. Not convinced it's a good default behavior.

So what would you have the default behaviour be?  

Not sure I agree that no cache scopes means private thing that should never be cached - if the user turned on caching, then presumably they do want caching.  We're just making sure that all cached data doesn't become 100% public if caching is turned on and they're using authentication.  It seems reasonable to me...

Reply:
>Also bear in mind that ExecutionAuth isn't just tokens - if it's public auth then that goes in there too.

Ah right, missed that.

> Not sure I agree that no cache scopes means private thing that should never be cached - if the user turned on caching, then presumably they do want caching. We're just making sure that all cached data doesn't become 100% public if caching is turned on and they're using authentication. It seems reasonable to me...

Hum fair point, never mind